Mathlib/Data/Finset/Density.lean

```diff
@@ -0,0 +1,215 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Algebra.Order.Field.Basic
+import Mathlib.Tactic.Positivity.Basic
+import Mathlib.Tactic.Positivity.Finset
+
+/-!
+# Density of a finite set
+
+This defines the density of a `Finset` and provides induction principles for finsets.
+
+## Main declarations
+
+* `Finset.dens t`: `dens s : ℕ` returns the density of `s : Finset α`.
+
+## Notation
+
+If you need to specify the field the density is valued in, `dens[𝕜] s` is notation for
+`(dens s : 𝕜)`. Note that no dot notation is allowed.
+-/
+
+open Function Multiset Nat
+
+variable {𝕜 α β : Type*} [Fintype α]
+
+namespace Finset
+section Semifield
+variable [Semifield 𝕜] {s t : Finset α} {a b : α}
+
+/-- `dens s` is the number of elements of `s`, aka its density. -/
+@[pp_dot] def dens (s : Finset α) : 𝕜 := s.card / Fintype.card α
```

Comment:
I'd argue that this should live in `NNRat` not `Semifield 𝕜`, for the same reason that `card` lives in `Nat` not `Semiring R`.

Reply:
... which maybe explains why you want `NNRat.cast`! Is `Rat` suitable enough in the meantime?

Reply:
Actually, the reason I want `NNRat.cast` is not this but the closely-related `Finset.expect` that still lives [in LeanAPAP](https://github.com/YaelDillies/LeanAPAP/blob/master/LeanAPAP/Prereqs/Expect/Basic.lean).

Having this be `NNRat`/`Rat`-valued will be very annoying since `dens` is meant to be an analytic object. Most uses of it will be in `Real`.

Reply:
Your concerns only apply to the case where we try to talk about the density valued across several types. I conjecture this barely ever happens as all that matters are the formulas for the density of a finset constructed in a given way.

Instead I am trying to optimise for those rewritings. Eg my API can rewrite `(s ∪ t).dens + (s ∩ t).dens = s.dens + t.dens` where everything is real-valued in one lemma. In contrast, your suggested API will painfully get there by doing
```
rw [← Nat.cast_add, dens_union_add_dens_inter, Nat.cast_add]
```
(assuming the rewrites by `Nat.cast` are unambiguous!)

Reply:
No, you wouldn't have to do that. With my suggestion, you would still have a theorem called `dens_union_add_dens_inter` that states that `dens[𝕜] (s ∪ t)+ dens[𝕜] (s ∩ t)= dens[𝕜] s+ dens[𝕜] t`, and you could rewrite using that theorem in one step. It's just that `dens[𝕜] s` would mean `(dens s : 𝕜) = ((s.card / Fintype.card α : ℚ) : 𝕜)` rather than meaning `(s.card : 𝕜) / (Fintype.card α : 𝕜)` like it currently does.

Reply:
Suppose you were to prove that if `s : Fin n` satisfies `dens[ℝ] s > exp(-c * (log n) ^ (1/9)))`, then `s` has a three-term arithmetic progression.

And then suppose that someone else wants to use your theorem on a set `s`. However, they only know `dens[ℚ] s > 1 / 100` because they did all their calculations with `dens[ℚ]`. Of course, this hypothetical person should be able to use your theorem, because `dens[ℚ] s` is equal to `dens[ℝ] s`. But currently, it's really hard for them to use it, because it's not obvious that they're equal.

(If you would say "Well, you should've used `dens[ℝ]` in all your calculations, because that's the standard", then that essentially means that `dens[ℝ]` is the only supported use case. In this case, `dens` should just be valued in `ℝ`, and you wouldn't even have to write the `[ℝ]`. Actually, I think this wouldn't be so bad.)

Reply:
I don't really think that's a strong argument. In your example, translation happens at exactly one point, so there's little overhead (and in contrast you aren't mentioning the countless uses of `NNRat.cast_add` that having `Finset.density` be `NNRat`-valued involves).

If you want an argument that I actually believe, here it is: `Finset.card` is `Nat`-valued, which in particular means it doesn't rely on the algebraic hierarchy.

Reply:
I think you do not understand what I am saying. You would not need to use `NNRat.cast_add` even a single time if the API is set up correctly.

Reply:
You're right. I had missed your point about just changing the definition.

If I understand you correctly, you now need two copies of each lemma:
* one for general semifields, stated with `NNRat.cast`
* one for `NNRat`, stated without `NNRat.cast`

What you seem to be missing is that the second class of lemmas can't be substituted by the first class of lemmas in rewrites. That's because `NNRat.cast : NNRat -> NNRat` is not syntactically the same as `id`.

Reply:
I have done the change.